File: нарешка/РОИ/mem_manager.cpp
#include <iostream>
#include <algorithm>
#include <vector>

int main(){
    int n, m;
    std::cin >> n >> m;
    std::vector<std::vector<int>>mem = std::vector<std::vector<int>>();
    mem.push_back({0, 0, 0});
    for(int i = 0; i < m; i++){
        int a;
        std::cin >> a;
        if(a >= 0){
            int index = -1;
            for(int j = 0; j < mem.size() - 1; j++){
                int endIndex = mem[j][0] + mem[j][1];
                int startIndex = mem[j + 1][0];
                if(startIndex - endIndex >= a){
                    index = endIndex;
                }
            }
            if(index == -1){
                index = mem[mem.size() - 1][0] + mem[mem.size() - 1][1];
            }
            if(index + a > n){
                std::cout << -1 << std::endl;
            }
            else{
                std::cout << index + 1 << std::endl;
                mem.push_back({index, a, i + 1});
                std::sort(mem.begin(), mem.end());
            }
        }
        else{
            int index = -1;
            for(int j = 0; j < mem.size(); j++){
                if(mem[j][2] == -a){
                    index = j;
                    break;
                }
            }
            if(index != -1){
                mem.erase(mem.begin() + index);
            }
        }
        /*for(int j = 0; j < mem.size(); j++){
            std::cout << mem[j][0] << " " << mem[j][1] << " " << mem[j][2] << std::endl;
        }*/
    }
    return 0;
}

File: нарешка/РОИ/mem_manager.py
n, m = map(int, input().split())

mem = [(0, 0, 0)]
#print(mem)

for i in range(m):
    a = int(input())
    if(a >= 0):
        index = -1
        for j in range(len(mem) - 1):
            endIndex = mem[j][0] + mem[j][1]
            startIndex = mem[j + 1][0]
            #print(startIndex - endIndex)
            if(startIndex - endIndex >= a):
                index = endIndex
                break
        if(index == -1):
            index = mem[len(mem) - 1][0] + mem[len(mem) - 1][1]

        if(index + a > n):
            print(-1)
        else:
            print(index + 1)
            mem.append((index, a, i + 1))
            mem.sort(key = lambda x: x[0])
    else:
        index = -1
        for j in range(len(mem)):
            if(mem[j][2] == -a):
                index = j
                break
        if(index != -1):mem.pop(index)
    #print(mem)
